### PR TITLE
Object based lighting system

### LINKED ISSUE #8063: Carpet smoke(or maybe smoke in general) leaves lightless turfs
Exactly what it says in the name, untill lights are reset by some means turfs occasionally stay unlit

### PR DESCRIPTION
**Pull request by MrPerson**
Uses actual objects on each non-space, dynamically lit turf. Light levels are switched back and forth via animate() and the object's alpha. Supporting colors shouldn't be too hard. Some hacky efficiency improvements means it isn't that much more expensive than current (I think, needs testing). Most of the lighting ss's cost is in checking all the lights and doing big loops, not anything actually in the loops themselves.

~~Start PDA flashlights on. This was to speed up testing but frankly I think it's a good change in general.~~

Added a Moved() proc. Called after a successful move.

In the future I hope to move off the luminosity var entirely but that was too slow in testing for me. That's what all that "for(area in sortedAreas) area.luminosity = 1" stuff in the lighting ss is, tests on removing luminosity outright.

Update: The image that was here wasn't really indicative of the way it looks. This is more what it looks like

testing_r

Also this now fixes #8063

### CODE REVIEW

**Comment by MrPerson** [comment]
I'm inclined to think this should go on an experimental branch and tested a lot more because I can imagine weird edge cases where it breaks things. Teleportation comes to mind.

**Comment by ACCount12** [comment]
I think it will increase client-server traffic, making every light change a lagspike for everyone with slow internet (i.e. me). Please check it.

**Comment by dannno** [comment]
woah, a big change being tested? BEFORE being implemented?
:+1:

(the lighting is cool btw)

**Comment by ACCount12** [comment]
And remove PDA flashlights on.

**Comment by Nienhaus** [comment]
It seems like the PDA light is deleyd movement in the gif.

**Comment by MrPerson** [comment]
I don't have my PDA on me in that gif, actually.

**Comment by Nienhaus** [comment]
The lamp then, sorry.

File: code/game/machinery/doors/brigdoors.dm [hunk 1]
```diff
@@ -42,15 +42,15 @@
 	pixel_y = ((src.dir & 3)? (src.dir ==1 ? 24 : -32) : (0))
 
 	spawn(20)
-		for(var/obj/machinery/door/window/brigdoor/M in world)
```
Comment (optimumtact) [inline code/game/machinery/doors/brigdoors.dm]: +500000

**Comment by WJohn** [comment]
How fast or slow is this compared to area lighting? How does it differ?

**Comment by MrPerson** [comment]
It takes about 30% longer for the lighting subsystem to initialize since it has to spawn more objects, but once it does the actual fire()'s seem to be about the same. Time spent calculating and actually rendering the objects is way smaller than keeping up with the light datums, which happens either way.

Back when I was creating an object for every single turf at roundstart, any for(/obj in world) loops would throw an infinite loop error. That might still happen, but if so I'll cry no tears about that "loss" of functionality.

Weak lights are very dark. Dunno how it differs with current, but I can make weaker lights a lot brighter artificially if people want. There's also very obvious and kinda ugly shadows all over the place which currently just get rounded together. I'll get more pictures to show what I mean but I'm still hoping someone hosts a test server. That's got more to do with mapping than anything else imo.

I'll remove the PDA lights unless there's a big outpouring of love for them. That can wait til another pr. Or never, whatever.

**Comment by Menshin** [comment]
Worth noting that this should prevent the luminosity var overflow (#7510, aka "dixel appearing from nowhere" or "gender-bender galore").

**Comment by MrPerson** [comment]
When I said initialization, I meant before the round starts. Someone on reddit was confused by this.

**Comment by RemieRichards** [comment]
So I ask, quizzically:
Why?
I mean you mentioned it wouldn't take much effort to add coloured lighting, but we can do that without object lighting too, I'm pretty sure.
You mention the speed is about the same at runtime but 30% worse at roundstart initialisation.

Is it even worth doing this?

**Comment by Menshin** [comment]
One reason may be that using the luminosity var limits us to a max luminosity of 6 (7 should be ok too).
That's only stated in the help file and there's no failsafe check, so there's items that outright ignores this limitation (mining stationary lamps iirc) and luminosity from different sources can add up to infinity, leading to overflow (and causing #7510).

**Comment by RemieRichards** [comment]
I mean yes the dixel is an issue, but this doesn't fix it, it simple circumvents it.

**Comment by Menshin** [comment]
We've no way to 'fix' the dixel (besides limiting luminosity to 6-7), it's actually Byond implementation that is 'at fault'.

**Comment by RemieRichards** [comment]
Yes menshin, I'm aware, and yes, Limiting luminosity to one below the byond limit is indeed the only "fix" we've got, so no, I wont count "fixing the dixel" as a pro for this PR, I just want to know why MrPerson is doing this since he usually doesn't do things on a whim.

**Comment by Miauw62** [comment]
This paves the way for colored lightning, obviously.
If i'm understanding this right, this could also allow people to create "antilights"?

And making PDA flashlights or flashlights in general darker is a good change.

**Comment by dannno** [comment]
Making PDA lights darker is ok-ish but don't ever remove them. I feel that they fill a good niche as an emergency light item

**Comment by Miauw62** [comment]
BTW, would people be OK with flashlights (including flashlights in helmets
but excluding PDA flashlights) having a battery that automatically
recharges when they're not in use?

**Comment by dannno** [comment]
I would, I don't doubt that there would be a lot of people against it though.

**Comment by Miauw62** [comment]
sssh i realized that this isnt the place for this discussion

**Comment by WJohn** [comment]
Honestly if this lets us have more than 7 levels of lighting I am MASSIVELY in favor of this. Currently lighting bleeds off extremely quickly, going from perfectly bright, to slightly less than that for the next 3, to neat instantly black after.

**Comment by ACCount12** [comment]
Remove pda flashlights starting on please.

**Comment by MrPerson** [comment]
I think I've capped it at 10 total luminosity = fully lit or basically 10 lighting levels plus the levels between, but that's a totally arbitrary number.

No, you really couldn't do colored lighting with the current system. Well I mean you could, but there'd be so many areas you'd basically ruin the advantages of using area-based lighting to begin with.

I think I have the times too short because it's hard to see the transition from light state to light state with that gif, but that to me is the big advantage instead of being instant. The idea I was shooting for was something like this:

testing

Course that was way too slow but I've not quite been able to replicate that yet.

**Comment by PKPenguin321** [comment]
that looks so cool

**Comment by WJohn** [comment]
I'm totally for having 10 levels of lighting if it means smoother transitions :+1:

**Comment by optimumtact** [comment]
I'm fairly certain byond added some sanity to the gender/lumen vars, I'm not sure if we're running that release howerver

**Comment by MrPerson** [comment]
The original plan was to not use luminosity at all, but even without that bug it's still nice to be able to have luminosity values of >6.

PDA flashlights start off again now. The code is still there to support having them on and I'll do a PR for it another day.

Still need to update some comments and get some live server testing.

Examples of what I mean by ugly lighting: 

ugly_lighting_1
ugly_lighting_2

**Comment by WJohn** [comment]
That doesn't feel like the lighting there is making any sense. It shouldn't immediately be dark on one side and not the other, unless light tubes are not as powerful as they used to be unless you stack them.

**Comment by optimumtact** [comment]
Was that comment about the old lighting or the new lighting, remember that our quadratic lighting implementation was removed in carn's ticker update, and we went back to the older lighting system

**Comment by WJohn** [comment]
It was about the pictures shown there, though I'll admit our current lighting is guilty of this too.

**Comment by MrPerson** [comment]
It's the same stuff, just more visually obvious. Currently you can have two squares that are 3.5 lighting levels apart and only display as being 1 level apart, depending on the exact numbers involved. Not sure why the turfs are coming out to the values they are, but maybe I can figure it out.

**Comment by optimumtact** [comment]
I still dont' understand what WJ is saying? Light is instantly gone because it's fast?

**Comment by MrPerson** [comment]
And here's with a tweak to make every light display at max brightness and fall off based on distance instead of having distance == brightness as well. Thus small lights are bright over a short distance instead of being weak over a short distance. Currently all lights are set to be as bright as possible but that could be changed pretty trivially another day.

ugly_lighting_1_r

ugly_lighting_2_r

**Comment by WJohn** [comment]
Yeah that's much better :+1:

**Comment by Miauw62** [comment]
PDA lights should really be weak over a short distance though (same for
computers)

**Comment by Petethegoat** [comment]
why objects instead of an image on the turf?

**Comment by MrPerson** [comment]
Images would have to be globally listed and outputted to every client which would make any future client.screen operation super expensive. Not to mention that testing by wb and anecdotes by Ter imply that images are super expensive. I hear BYOND checks every image in every client's screen every tick to see if that image changed. Or something along those lines.

And more importantly, I didn't think of images until you mentioned them.

**Comment by MrPerson** [comment]
I don't think it needs any more edits at this point beyond fixing bugs

**Comment by Petethegoat** [comment]
as an overlay? does the image expensiveness expand to that do you know?

if so then nvm but i'm mainly just curious for my own use

**Comment by MrPerson** [comment]
You can't animate() an overlay

**Comment by Petethegoat** [comment]
ah, neat. so it'd be fine if you didn't require the animation? neat.

**Comment by optimumtact** [comment]
:goat:

**Comment by d3athrow** [comment]
Its neat but I feel as if its a bit too slow for no real benefit, it also poses the issue of removing the ability for people to play if they aren't able to use hardware acceleration.

>You can't animate() an overlay

You can its just a pain in the ass to get it to work.

**Comment by MrPerson** [comment]
Slow? Processing wise it's going to be about the same. The speed loss from lighting is from doing big loops constantly, not from anything in the big loops. shift_to_subarea() (current) and redraw_lighting() (in pr) are about the same, speed-wise. They don't get called that often, quite frankly.

I guess I should probably figure out a way to allow colors and not use BLEND_MULTIPLY, if possible. Quite frankly though, I don't care much about the 1 out of 50 users that use software mode so if it winds up not being possible, I'm not going to shed any tears.

**Comment by MrPerson** [comment]
Current code supports software mode 100%. Color + software mode support isn't something I know how to do offhand.

**Comment by An0n3** [comment]
Colored lighting looks disgusting on other servers pls no.  This looks neat, but keep it a uniform color.

**Comment by MrPerson** [comment]
Should be ready to merge again. Sorry about making it impossible to diff but it's better to be in a module.

**Comment by Miauw62** [comment]
here goes nothing

**Comment by Xhuis** [comment]
;_;7

**Comment by Crystalwarrior** [comment]
So, looking through the code, I noticed that the way it's handled allows for a much easier implementation of colored lighting system.
Are you going to try and whip up a prototype of that anytime soon? I really want to see how you'd do it.

**Pull request merged**